File: Source/Triggers/RestartGameTrigger.cs
using Celeste.Mod.Entities;
using Microsoft.Xna.Framework;

namespace Celeste.Mod.KoseiHelper.Triggers;

[CustomEntity("KoseiHelper/RestartGameTrigger")]
public class RestartGameTrigger : Trigger
{
    public bool restartOnlyLevel = false;
    public TriggerMode triggerMode;
    public string flag;

    public RestartGameTrigger(EntityData data, Vector2 offset) : base(data, offset)
    {
        restartOnlyLevel = data.Bool("restartOnlyLevel", false);
        triggerMode = data.Enum("triggerMode", TriggerMode.OnEnter);
        flag = data.Attr("flag", "");
    }
    public override void OnEnter(Player player)
    {
        base.OnEnter(player);
        Level level = SceneAs<Level>();
        if (triggerMode == TriggerMode.OnEnter && (string.IsNullOrEmpty(flag) || level.Session.GetFlag(flag)))
        {
            Restart();
        }
    }

    public override void OnLeave(Player player)
    {
        base.OnEnter(player);
        Level level = SceneAs<Level>();
        if (triggerMode == TriggerMode.OnLeave && (string.IsNullOrEmpty(flag) || level.Session.GetFlag(flag)))
        {
            Restart();
        }
    }

    public override void OnStay(Player player)
    {
        base.OnEnter(player);
        Level level = SceneAs<Level>();
        if (triggerMode == TriggerMode.OnStay && (string.IsNullOrEmpty(flag) || level.Session.GetFlag(flag)))
        {
            Restart();
        }
    }

    public void Restart()
    {
        Level level = SceneAs<Level>();
        if (restartOnlyLevel)
            AreaData.Areas[level.Session.Area.ID].Mode[(int)level.Session.Area.Mode].MapData.Reload();
        else
            Everest.QuickFullRestart();
        RemoveSelf();
    }
}

File: Loenn/triggers/RestartGameTrigger.lua
local RestartGameTrigger = {}

RestartGameTrigger.name = "KoseiHelper/RestartGameTrigger"
RestartGameTrigger.placements = {
    name = "RestartGameTrigger",
    data = {
		triggerMode = "OnEnter",
		flag = "",
		restartOnlyLevel = false,
		onlyOnce = false
		--delay = 0
    }
}

RestartGameTrigger.fieldInformation = function (entity) return {
	triggerMode = {
		options = {
		"OnEnter",
		"OnLeave",
		"OnStay"
		},
		editable = false
	},
	delay = {
		minimumValue = 0
	}
}
end

return RestartGameTrigger